src/dune_rules/melange_rules.ml

```diff
@@ -34,46 +34,71 @@ let local_of_lib ~loc lib =
           (Lib_name.to_string lib_name)
       ]
 
-let js_includes ~loc ~sctx ~target_dir ~requires_link ~scope ~js_ext =
-  let open Resolve.Memo.O in
-  Command.Args.memo
-    (Resolve.Memo.args
-       (let* (libs : Lib.t list) = requires_link in
-        let project = Scope.project scope in
-        let deps_of_lib (lib : Lib.t) =
-          let dst_dir =
-            let lib_dir =
-              let info = local_of_lib ~loc lib |> Lib.Local.info in
-              Lib_info.src_dir info
-            in
-            lib_output_dir ~target_dir ~lib_dir
-          in
-          let open Memo.O in
-          let* source_modules =
-            Dir_contents.modules_of_lib sctx lib
-            >>| Option.value_exn >>| Modules.impl_only
-          in
-          let of_module m =
-            let output = make_js_name ~js_ext ~dst_dir m in
-            Dep.file (Path.build output)
-          in
-          Resolve.Memo.return
-            (List.map source_modules ~f:of_module |> Dep.Set.of_list)
+let virtual_dst_dir_and_modules sctx lib ~loc ~target_dir =
+  match Lib.implements lib with
+  | None -> Resolve.Memo.return None
+  | Some vlib ->
+    let open Resolve.Memo.O in
+    let* vlib = vlib in
+    let info =
+      let vlib = local_of_lib ~loc vlib in
+      Lib.Local.info vlib
+    in
+    let+ modules =
+      Dir_contents.modules_of_lib sctx vlib
+      |> Resolve.Memo.lift_memo >>| Option.value_exn
+    in
+    let lib_dir = Lib_info.src_dir info in
+    let dst_dir = lib_output_dir ~target_dir ~lib_dir in
+    Some (dst_dir, vlib, modules)
+
+let js_includes ~loc ~sctx ~target_dir ~(requires_link : Lib.t list Resolve.t)
+    ~scope ~js_ext =
+  let project = Scope.project scope in
+  let deps_of_lib =
+    let of_module m ~dst_dir = make_js_name ~js_ext ~dst_dir m |> Path.build in
+    fun lib ->
+      let lib_dir =
+        local_of_lib ~loc lib |> Lib.Local.info |> Lib_info.src_dir
+      in
+      let open Memo.O in
+      let* source_modules =
+        Dir_contents.modules_of_lib sctx lib
+        >>| Option.value_exn >>| Modules.impl_only
+      in
+      let open Resolve.Memo.O in
+      let+ virtual_deps =
+        let+ dst_dir_and_virtual_modules =
+          virtual_dst_dir_and_modules sctx lib ~loc ~target_dir
         in
-        let* hidden_libs = Resolve.Memo.List.map libs ~f:deps_of_lib in
-        let hidden_deps = Dep.Set.union_all hidden_libs in
-        Resolve.Memo.return
-          (Command.Args.S
-             [ Lib_flags.L.include_flags ~project libs Melange
-             ; Hidden_deps hidden_deps
-             ])))
+        match dst_dir_and_virtual_modules with
+        | None -> []
+        | Some (dst_dir, _vlib, virtual_modules) ->
+          let source_modules = Modules.impl_only virtual_modules in
```

Comment:
In https://github.com/ocaml/dune/pull/6495#discussion_r1028780498, you mentioned:

> Use `Modules.split_by_lib` to get the modules for the virtual library and implementation separately

I understand it's not necessary as `virtual_dst_dir_and_modules` is calling `Dir_contents.modules_of_lib sctx vlib` already?

Reply:
Oops, it's still necessary. I added it now.

Reply:
Btw, I misdiagnosed the issue originally. The issue was that we'd setup the same rules but in different directories. E.g, we'd have:

```
vlib/foo.ml
impl/foo.ml
```

Which is unnecessary.